### ISSUE TITLE
[regression] "th" no longer works as ordinal specifier in date specifications

### ISSUE THREAD

**Issue by stephane-chazelas** [ksh93/ksh#733]
(from https://unix.stackexchange.com/questions/486656/pick-off-first-dates-using-cal-function-in-bash/773307#773307)

```
$ printf '%(%F)T\n' {first,second,third,fourth,fifth,1st,2nd,3rd,4th,5th}' tuesday in march 2016'
2016-03-01
2016-03-08
2016-03-15
2016-03-22
2016-03-29
2016-03-01
2016-03-08
2016-03-15
2016-04-06 (!)
2016-03-14 (!)
```

Worked fine in ksh93u ksh93u+ and the 93u+m/1.0.0-beta.2 2021-12-17 in Ubuntu 22.04. Already broken in the 93u+m/1.0.8 2024-01-01 currently in Debian.

**Comment by stephane-chazelas** [comment]
See also:

```
$ ksh -c 'printf "%(%F)T\n" "4nd tuesday in march 2016"' '
2016-03-08
```

Which changed from 2016-03-22, that is where `4nd` is interpreted as `2nd` instead of `4th`.

IMO, those `st`, `nd`, `rd`, `th` suffixes should only be interpreted as ordinal specifiers rather than carry numeric information. It's important to be able to use `${n}th` without having to adapt the `th` to `st`, `nd`, `rd` depending on the value of `$n` like in:

```
year=2016 month=3 first=1 last=5 day=tuesday
printf '%(%F)T\n' {"$first".."$last"}"th  $day in $year-$month"
```

Which used to give (with any of `th`, `nd`, `st`...):

```
2016-03-01
2016-03-08
2016-03-15
2016-03-22
2016-03-29
```

But now give:

```
2016-03-16
2016-03-23
2016-03-30
2016-04-06
2016-03-14
```

With `th` (when run on 2024-03-28 at least) and:

```
2016-03-08
2016-03-08
2016-03-08
2016-03-08
2016-03-08
```

With `th` replaced with `nd`.

**Comment by stephane-chazelas** [comment]
Looks like it was broken by 1a9d8adaeb4415a0239b463eab73fe51239a26eb from pull request #641 which was addressing issue #182

**Comment by McDutchie** [comment]
Ping @phidebian -- any ideas?

**Comment by phidebian** [comment]
Will try to look tonite...

**Comment by phidebian** [comment]
It was a left over, while I reworked this things some time ago, I normalised the 'f' variable usage (specially the +1 -1 usage), and the criptic TM_ORDINAL case was kinda magical and I left it untouched.
Now thanx @stephane-chazelas, I got a test case and can understand it.
I added a test case to catch any regression in this area if one try to make printf better :-)

**Comment by McDutchie** [comment]
Fix pending in #734. My initial tests confirm that all of Stéphane's reproducers are fixed. Tomorrow I'll add the requisite housekeeping and merge it. Thanks!
